src/sentry/models/groupsnooze.py

```diff
@@ -142,6 +142,53 @@ def test_user_rates(self):
 
         return True
 
+    def test_user_counts(self, group: Group | None = None) -> bool:
+        """
+        Test if the number of unique users seen by the group is below the snooze threshold.
+        Returns: True if below threshold, False otherwise.
+
+        - Non-cached version of the function queries Snuba for the real count every time.
+        - Cached version uses counters to store the number of events seen since last check,
+          if it's less than the number of users needed to reach the threshold, we can be sure
+          that we couldn't have reach enough users to reach the threshold, so there's no need
+          to query Snuba. This functionality relies on the fact that this is called in
+          post-processing for every event, so we can assume that the call-count == event count.
+        """
+        if features.has(
+            "organizations:groupsnooze-cached-counts", organization=self.group.project.organization
+        ):
+            return self.test_user_counts_w_cache(group)
+        else:
+            return self.test_user_counts_no_cache(group)
+
+    def test_user_counts_no_cache(self, group=None):
+        metrics.incr("groupsnooze.test_user_counts")
+        metrics.incr("groupsnooze.test_user_counts.real_count")
+
+        threshold = self.user_count + self.state["users_seen"]
+        real_count = group.count_users_seen(
+            referrer=Referrer.TAGSTORE_GET_GROUPS_USER_COUNTS_GROUP_SNOOZE.value
+        )
+        return real_count < threshold
+
+    def test_user_counts_w_cache(self, group=None):
+        metrics.incr("groupsnooze.test_user_counts", tags={"cached": "true"})
+        cache_key = f"groupsnooze:v1:{self.id}:test_user_counts:events_seen_counter"
+
+        threshold = self.user_count + self.state["users_seen"]
+
+        if cache.get(cache_key, float("inf")) < threshold - 1:
+            # if we've seen less than that many events, we can't possibly have seen enough users
+            cache.incr(cache_key)
```

Comment:
I think that we should also look at the post `incr` value that is returned, and if that's above `threshold-1` then we query. This just avoids a few race conditions where two processes fetch the same value, both increment, and one ends up over the threshold

Reply:
good callout, I've changed it to increase and get value from that before the if check